CIP-XXXX/CIP-XXXX.md

```diff
@@ -0,0 +1,558 @@
+
+### Operation semantics
+
+Our proposed operations will have the following signatures:
+
+* ``bitwiseShift :: BuiltinByteString -> BuiltinInteger -> BuiltinByteString``
+* ``bitwiseRotate :: BuiltinByteString -> BuiltinInteger -> BuiltinByteString``
+* ``countSetBits :: BuiltinByteString -> BuiltinInteger``
+* ``findFirstSetBit :: BuiltinByteString -> BuiltinInteger``
+
+We assume the following costing, for both memory and execution time:
+
+| Operation | Execution time cost | Memory cost |
+|-----------|------|
+|`bitwiseShift`| Linear in the `BuiltinByteString` argument | As execution time
+|
+|`bitwiseRotate` | Linear in the `BuiltinByteString` argument | As execution
+time |
+|`countSetBits` | Linear in the argument | Constant |
+|`findFirstSetBit` | Linear in the argument | Constant |
+
+#### `bitwiseShift`
+
+`bitwiseShift` takes two arguments; we name and describe them below.
+
+1. The `BuiltinByteString` to be shifted. This is the _data argument_.
+2. The shift, whose sign indicates direction and whose magnitude indicates the
+   size of the shift. This is the _shift argument_, and has type
+   `BuiltinInteger`.
+
+Let $b$ refer to the data argument, whose length in bytes is $n$, and let $i$
+refer to the shift argument. Let the result of `bitwiseShift` called with $b$
+and $i$ be $b_r$, also of length $n$. 
+
+For all $j \in 0, 1, \ldots 8 \cdot n - 1$, we have
+
+$$
+b_r[j] = \begin{cases}
+         b[j - i] & \text{if } j - i \in 0, 1, \ldots 8 \cdot n - 1\\
+         0 & \text{otherwise} \\
+         \end{cases}
+$$
+
+Some examples of the intended behaviour of `bitwiseShift` follow. For
+brevity, we write `BuiltinByteString` literals as lists of hexadecimal values.
+
+```
+-- Shifting the empty bytestring does nothing
+bitwiseShift [] 3 => []
+-- Regardless of direction
+bitwiseShift [] (-3) => []
+-- Positive shifts move bits to higher indexes, cutting off high indexes and
+-- filling low ones with zeroes
+bitwiseShift [0xEB, 0xFC] 5 => [0x7F, 0x80]
+-- Negative shifts move bits to lower indexes, cutting off low indexes and
+-- filling high ones with zeroes
+bitwiseShift [0xEB, 0xFC] (-5) => [0x07, 0x5F]
+-- Shifting by the total number of bits or more clears all bytes
+bitwiseShift [0xEB, 0xFC] 16 => [0x00, 0x00]
+-- Regardless of direction
+bitwiseShift [0xEB, 0xFC] (-16) => [0x00, 0x00]
+```
+
+#### `bitwiseRotate`
+
+`bitwiseRotate` takes two arguments; we name and describe them below.
+
+1. The `BuiltinByteString` to be rotated. This is the _data argument_.
+2. The rotation, whose sign indicates direction and whose magnitude indicates 
+   the size of the rotation. This is the _rotation argument_, and has type
+   `BuiltinInteger`.
+
+Let $b$ refer to the data argument, whose length in bytes is $n$, and let $i$
+refer to the rotation argument. Let the result of `bitwiseRotate` called with $b$
+and $i$ be $b_r$, also of length $n$. 
+
+For all $j \in 0, 1, \ldots 8 \cdot n - 1$, we have $b_r = b[j - i \mod 8 \cdot n]$.
```

Comment:
There's a rendering problem here too!

![Notsupported](https://github.com/cardano-foundation/CIPs/assets/2124565/1978313c-39cd-4a3a-837f-95acd08e5907)

Reply:
That is _extremely_ odd. Will investigate: probably a forgotten delimiter somewhere.

Reply:
That was... not sure what that was, but it's fixed now.